libs/elasticsearch-core/src/main/java/org/elasticsearch/core/internal/io/IOUtils.java

```diff
@@ -0,0 +1,291 @@
+                b.append("   ")
+                        .append(kv.getKey().toAbsolutePath())
+                        .append(": ")
+                        .append(kv.getValue())
+                        .append("\n");
+            }
+            throw new IOException(b.toString());
+        }
+    }
+
+    private static LinkedHashMap<Path,Throwable> rm(final LinkedHashMap<Path,Throwable> unremoved, final Path... locations) {
+        if (locations != null) {
+            for (final Path location : locations) {
+                // TODO: remove this leniency
+                if (location != null && Files.exists(location)) {
+                    try {
+                        Files.walkFileTree(location, new FileVisitor<Path>() {
+                            @Override
+                            public FileVisitResult preVisitDirectory(final Path dir, final BasicFileAttributes attrs) throws IOException {
+                                return FileVisitResult.CONTINUE;
+                            }
+
+                            @Override
+                            public FileVisitResult postVisitDirectory(final Path dir, final IOException impossible) throws IOException {
+                                assert impossible == null;
+
+                                try {
+                                    Files.delete(dir);
+                                } catch (final IOException e) {
+                                    unremoved.put(dir, e);
+                                }
+                                return FileVisitResult.CONTINUE;
+                            }
+
+                            @Override
+                            public FileVisitResult visitFile(final Path file, final BasicFileAttributes attrs) throws IOException {
+                                try {
+                                    Files.delete(file);
+                                } catch (final IOException exc) {
+                                    unremoved.put(file, exc);
+                                }
+                                return FileVisitResult.CONTINUE;
+                            }
+
+                            @Override
+                            public FileVisitResult visitFileFailed(final Path file, final IOException exc) throws IOException {
+                                if (exc != null) {
+                                    unremoved.put(file, exc);
+                                }
+                                return FileVisitResult.CONTINUE;
+                            }
+                        });
+                    } catch (final IOException impossible) {
+                        throw new AssertionError("visitor threw exception", impossible);
+                    }
+                }
+            }
+        }
+        return unremoved;
+    }
+
+    // TODO: replace with constants class if needed (cf. org.apache.lucene.util.Constants)
```

Comment:
I put this comment here 😐. In Lucene, these use `org.apache.lucene.util.Constants` which we do not have access to here. I did not want to wholesale copy right now (and would like to avoid if reasonable).